src/cellarr/dataloader.py

```diff
@@ -1,15 +1,374 @@
-# config["sm.compute_concurrency_level"] = 1
-# config["sm.io_concurrency_level"] = 1
-# config["sm.num_async_threads"] = 1
-# config["sm.num_reader_threads"] = 1
-# config["sm.num_tbb_threads"] = 1
-# config["sm.num_writer_threads"] = 1
+config["sm.compute_concurrency_level"] = 1
+config["sm.io_concurrency_level"] = 1
+config["sm.num_async_threads"] = 1
+config["sm.num_reader_threads"] = 1
+config["sm.num_tbb_threads"] = 1
+config["sm.num_writer_threads"] = 1
 config["vfs.num_threads"] = 1
+
+
+class scDataset(Dataset):
+    """A pytorch Dataset that enumerates cells and cell labels in TileDB."""
+
+    def __init__(
+        self,
+        data_df: "pandas.DataFrame",
+        matrix_tdb: "tiledb.libtiledb.SparseArrayImpl",
+        matrix_shape: tuple,
+        gene_indices: List[int],
+        label_column: str,
+        study_column: str,
+        lognorm: bool = True,
+        target_sum: float = 1e4,
+    ):
+        """Constructor.
+
+        Parameters
+        ----------
+        data_df: pandas.DataFrame
+            Pandas dataframe of valid cells.
+        matrix_tdb: tiledb.libtiledb.SparseArrayImpl
+            Counts TileDB.
+        matrix_shape: tuple,
+            Shape of the counts matrix
+        gene_indices: List[int]
+            The index of genes to return.
+        label_column: str, default: "cellTypeName"
+            Label column name.
+        study_column: str, default: "dataset"
+            Study column name.
+        lognorm: bool, default: True
+            Whether to return log normalized expression instead of raw counts.
+        target_sum: float, default: 1e4
+            Target sum for log normalization.
+        """
+
+        self.data_df = data_df
+        self.matrix_tdb = matrix_tdb
+        self.matrix_shape = matrix_shape
+        self.gene_indices = gene_indices
+        self.lognorm = lognorm
+        self.target_sum = target_sum
+        self.label_column = label_column
+        self.study_column = study_column
+
+    def __len__(self):
+        return self.data_df.shape[0]
+
+    def __getitem__(self, idx):
+        # data, label, study
+        cell_idx = self.data_df.index[idx]
+        results = self.matrix_tdb.multi_index[cell_idx, :]
+        counts = coo_matrix(
+            (results["data"], (results["cell_index"], results["gene_index"])),
+            shape=self.matrix_shape,
+        ).tocsr()
+        counts = counts[cell_idx, :]
+        counts = counts[:, self.gene_indices]
+
+        X = counts.astype(np.float32)
+        if self.lognorm:
+            counts_per_cell = counts.sum(axis=1)
+            counts_per_cell = np.ravel(counts_per_cell)
+            counts_per_cell = counts_per_cell / self.target_sum
+            X = X / counts_per_cell[:, None]
+            X = csr_matrix(X).log1p()
+
+        X = X.toarray()
+
+        return (
+            X,
+            self.data_df.loc[cell_idx, self.label_column],
+            self.data_df.loc[cell_idx, self.study_column],
+        )
```

Comment:
hmm now i am thinking if this should just pass the function to the `CellArrDataset`, get the matrix, gene and cell dataframes, and only return what you need?

Reply:
I'm working off of the data_df frame though which has be filtered.